File: spring-boot-logging/src/test/java/x/loggy/data/SqlQueries.java
package x.loggy.data;

import ch.qos.logback.classic.Logger;
import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.core.AppenderBase;
import org.springframework.jdbc.core.JdbcTemplate;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import static ch.qos.logback.classic.Level.DEBUG;
import static java.util.regex.Pattern.compile;
import static org.slf4j.LoggerFactory.getLogger;

public final class SqlQueries
        extends AbstractList<String> {
    private static final Pattern queryOnly = compile(
            "^Executing prepared SQL statement \\[(.*)]$");
    private final List<String> queries = new ArrayList<>();
    @SuppressWarnings("unused")
    private final Appender appender = new Appender();

    @Override
    public int size() {
        return queries.size();
    }

    @Override
    public String get(final int index) {
        return queries.get(index);
    }

    @Override
    public void clear() {
        queries.clear();
    }

    private class Appender
            extends AppenderBase<ILoggingEvent> {
        private Appender() {
            final var logger = (Logger) getLogger(
                    JdbcTemplate.class.getName());
            logger.setLevel(DEBUG);
            // TODO: How to restore DEBUG console logging when done?
            logger.setAdditive(false); // Avoid extra DEBUG console logging
            logger.addAppender(this);
            start();
        }

        @Override
        protected void append(final ILoggingEvent eventObject) {
            final var message = eventObject.getMessage();
            final var matcher = queryOnly.matcher(message);
            if (!matcher.find()) return;
            queries.add(matcher.group(1));
        }
    }
}

File: domain-persistence-modeling/kotlin-spring-boot/src/main/kotlin/x/domainpersistencemodeling/SqlQueries.kt
package x.domainpersistencemodeling

import ch.qos.logback.classic.Level.DEBUG
import ch.qos.logback.classic.Logger
import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.core.AppenderBase
import net.sf.jsqlparser.JSQLParserException
import net.sf.jsqlparser.parser.CCJSqlParserUtil
import org.slf4j.LoggerFactory.getLogger
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component
import java.util.AbstractList
import java.util.ArrayList
import java.util.Locale
import java.util.regex.Pattern

@Component
class SqlQueries
    : AbstractList<String>(),
        AutoCloseable {
    private val queries: MutableList<String> = ArrayList()
    private val appender = Appender()

    override val size: Int
        get() = queries.size

    override fun get(index: Int): String {
        return queries[index]
    }

    override fun clear() {
        queries.clear()
    }

    override fun close() {
        appender.stop()
    }

    val byType: Map<String, List<String>>
        get() = groupBy { bucket(it) }

    private inner class Appender : AppenderBase<ILoggingEvent>() {
        override fun append(eventObject: ILoggingEvent) {
            val message = eventObject.message
            val matcher =
                    queryOnly.matcher(message)
            if (!matcher.find()) return
            val query = matcher.group(1)
            queries.add(query)
        }

        init {
            val logger = getLogger(
                    JdbcTemplate::class.java.name) as Logger
            logger.level = DEBUG
            // TODO: How to restore DEBUG console logging when done?
            logger.isAdditive = false // Avoid extra DEBUG console logging
            logger.addAppender(this)
            start()
        }
    }

    companion object {
        private val queryOnly = Pattern.compile(
                "^Executing prepared SQL statement \\[(.*)]$")
    }
}

fun bucket(query: String): String {
    return try {
        CCJSqlParserUtil.parse(query).javaClass.simpleName
                .replace("Statement", "")
                .toUpperCase(Locale.US)
    } catch (e: JSQLParserException) {
        "INVALID"
    }
}
